Fix for data spam for LootR

Speiger opened the pull request:
This is a fully functioning & tested fix for lootRs data storage.

What did I change?
LootR now stores everything in independent files in the player folder that only has 1-2 files per player.
Since everything (except decay) is based around player actions, we can save everything inside the player.
If the player logs off the file is automatically unloaded with it.
If you worry about memory leaks this is exactly the same amount of memory usage as before.
Just a cleaner file system and some performance improvements thanks to the better dealing of "timed data".

Also this includes a fully functioning migration service that automigrates everything on worldload.

I hope this helps.

Changelog:
-Fixed: Spamming of Files that don't use any space and will never unload once loaded until server restart.
-Added: LootR data is now saved in player specific files when the player logs off the files will be unloaded.
-Fixed: If to many chests are in the Decay/Refresh queue they could lag. Now using a better system to do the same job. (Works even if players are logged off)
-Added: Migration System that automatically converts old system into the new System.
-Fixed: Cleaning of Player Specfic data is now 1000x faster because the data is now mapped per player instead of per chest/minecart.

Speiger committed:
Reworked data commit.

-Fixed: Spamming of Files that don't use any space and will never unload once loaded until server restart.
-Added: LootR data is now saved in player specific files when the player logs off the files will be unloaded.
-Fixed: If to many chests are in the Decay/Refresh queue they could lag. Now using a better system to do the same job. (Works even if players are logged off)
-Added: Migration System that automatically converts old system into the new System.
-Fixed: Cleaning of Player Specfic data is now 1000x faster because the data is now mapped per player instead of per chest/minecart.

noobanidus commented:
Thanks for the PR! I'll take a look over it some time this week.

noobanidus commented:
So, I appreciate the effort you've put into this, and it's very obvious to me that you're quite concerned about the way the system currently works.

For the record, can you clarify exactly *what* your concerns are with the current system? You've mentioned that a number of files are created and that these cached objects are held in memory -- do you have a memory dump to indicate that there's a significant build-up? Or do you simply object to the number of files on a personal level? What exactly is it about this that's problematic?

You've laid out a number of concerns including memory leaks, small files being kept in memory, etc, but to commit to wholly changing things in this fashion, I need some evidence that it's going to dramatically improve something. That is, a "before" and "after" snapshot.

I'm very reluctant to wholly change things in this fashion as it means I'm responsible for maintaining it in the future, as well as responsible for porting it to later versions. Given that I'm currently maintaining the 1.16.5 and 1.18 versions, a divergence between them is going to be frustrating.

Speiger commented:
> For the record, can you clarify exactly what your concerns are with the current system? You've mentioned that a number of files are created and that these cached objects are held in memory -- do you have a memory dump to indicate that there's a significant build-up? Or do you simply object to the number of files on a personal level? What exactly is it about this that's problematic?

The main problem is that you spam files for every single chest that has been interacted with, which unless there is a really good reason for it is really really bad practice. 
Especially for "Player Specific Data".
To give you a example
![image](https://user-images.githubusercontent.com/2556269/154846502-c95c88f3-c3b0-4d54-87fc-bce9a031fe40.png)
File systems don't really like thousands of files being in one folder. If actually lags out quite a bit if you try to do that.
And your system has the potential to do such a thing really likely (i mean we  are already at almost a 1k files here)
And it doesn't help that you throw all dimensions into 1 folder.

So what I am concerned about is that you throw thousands of small files into 1 folder where you cant really deal with them because there is no real identifiable way for the user to distinguish them.

> You've laid out a number of concerns including memory leaks, small files being kept in memory, etc, but to commit to wholly changing things in this fashion, I need some evidence that it's going to dramatically improve something. That is, a "before" and "after" snapshot.

Yeah if you have a million instances open at the same time it can result into a memory leak but this is only my "Insane optimazation head that warns here". Though while this has more files open at once if the server frequently restarts and exploring players are always online it still allows data to unload and also to trace things back more easily.
If a server doesn't restart yours has thousands of files in ram that only exists so "IF" a player (even if all players are logged out) might interact with them. 

So really bad practice + wastefull system that just waits on triggers without waiting.

> I'm very reluctant to wholly change things in this fashion as it means I'm responsible for maintaining it in the future, as well as responsible for porting it to later versions. Given that I'm currently maintaining the 1.16.5 and 1.18 versions, a divergence between them is going to be frustrating.

You don't have to accept this patch, but this is for me a bad enough experience having to deal with lootrs spam a file where I am saying I am never reinstalling that version ever again. And throwing a warn out, what people get into when they want to add this to multiplayer servers.

Its the reason why I invested 5 hours designing a patch that adresses everything and will improve performance for stuff like decay/refreshes.

About maintainablity.

This is as maintainable as your old system. Because it uses VERY ESSENTIAL events that will exist in next versions for sure.
And even if they don't exist *cough*fabric*cough* these events are so easily recreate able via mixins that the overhead would be only the initial setup and after that is done you no longer have to worry about these.

> EDIT: And also, as I believe I mentioned previously, I'm currently going through a period of very high stress. Unless there is something dramatically wrong (such as the 1.16.5 version crashing with the null worlds if the ticking happens after a dimension is unloaded), I'm unwilling to change things so significantly.

That is your choice, I am going to switch to my patch for my server because I honestly don't want to deal with a messy file system anymore.

Speiger commented:
And Migrated our server with the custom version.
From +500 files down to 1 file per player. (14 files for me, 28 with corruption protection)
And if players no longer show up i can just delete all the player specific data if desired.

![image](https://user-images.githubusercontent.com/2556269/154852156-c640b38c-3ef4-4b8b-94e1-a4f42f575c1b.png)

Speiger commented:
Ok the patch is still buggy. I need to test it a bit more.
So I wouldn't integrate it right now. But I will fix the bugs in the coming days.

Speiger committed:
Fixed bugs that were found during testing.

-Fixed: Refresh/Decay was kicking in. Even so it was disabled.
-Fixed: LootChests would loose their data after single use/world unload.
-Fixed: LootChests not loading their player data.

Speiger commented:
And fixed all the bugs that were found.
I will test it on my server where we play on. To make sure we don't introduce bugs.

noobanidus commented:
> To give you a example ![image](https://user-images.githubusercontent.com/2556269/154846502-c95c88f3-c3b0-4d54-87fc-bce9a031fe40.png)

You're absolutely right, I can see how this would make it an issue if someone were attempting to interact with other files in the data folder.

Thanks for giving me this bit of info, it's completely changed my mind on the whole issue.

I do have an idea for an optimisation with regards to how to deal with this that doesn't necessarily involve such a major overhaul, involving sorting into subdirectories. I'll look into these options but I'll also look over your PR again.

Speiger commented:
\o/ Finally.

My patch is now running on our test server (12 people) without issues.

It is not a giant overhaul but it is just changing the system from a "PerBlockSystem" to a "PerPlayerSystem"
which since everything (Except decay) is designed to be "PerPlayer" makes a lot of sense.

Also if players leave. You can just delete their player files from the file system and everything would be fine (since the data is unloaded the moment the player leaves)
Really good on maintainability.

noobanidus commented:
As always, my primary concern is inflation of specific files resulting in corruption of data. Having a per-block system means that, even if that file becomes over-inflated, it is just *that chest* that gets corrupted rather than everything. Likewise, refreshing chests need to clear the contents of the **entire** chest for everyone. 

On the other hand, storing everything per-player has the chance that, if the player data file becomes corrupted, the *entirety* of that is now gone.

I'm currently working on a patch to subdivide using the first 2 letters of the UUID, thus:

```
data/
  Lootr/
    Lootr-Advancement.dat
    ... etc
    6b/Lootr-chests-overworld-6bc3241b-96af-476c-a1f9-9c928834fc35.dat
    6e/Lootr-chests-overworld-6ef8042a-c0fb-43d3-9126-1c70a64d5b89.dat
```

This will result in a cleaner directory, and prevent any file system issues of having too many files in the one directory.

It also means that my eventual plan of an "Uninstaller" mod is still possible, as, while the block entity & block entity data will be removed once the mod is removed, the relevant saved data files can be used to reconstruct a Vanilla chest in the correct location.

Speiger commented:
> As always, my primary concern is inflation of specific files resulting in corruption of data. Having a per-block system means that, even if that file becomes over-inflated, it is just that chest that gets corrupted rather than everything. Likewise, refreshing chests need to clear the contents of the entire chest for everyone.

Actually you didn't fully read my code.
Since i copied the saving system from mcs PlayerData Saving System it has a build in Corruption detection/prevention.

It saves into a temp file then saves everything into there.
After that it tires to move the previously saved data into the "backup file" if that fails for 10 times (because of glitches or other issues) it interrupts the saving attempt.
If it is successful then it tries to save the  temp file as the previously saved one.
So it has a much much much higher Corruption protection then WorldSavedData

And it is proven to be working since that system is used across all minecraft servers that exist out there, including singleplayer instances.
And if your data corrupts randomly that info would spread like wildfire.

Also one downside of worldsaveddata is you can not delete it without shutting down the server, or doing it from the ram side.
